File: server/src/main/java/vn/edu/hust/volunteer_app/models/entity/Fanpage.java
package vn.edu.hust.volunteer_app.models.entity;

import java.sql.Date;

import jakarta.persistence.*;
import jakarta.validation.constraints.Size;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import vn.edu.hust.volunteer_app.annotation.ValidName;

@Entity
@Table(name = "fanpage")
@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
public class Fanpage {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private Integer id;

    @Column(name = "fanpage_name", unique = true)
    @Size(min = 5,max = 100)
    @ValidName
    private String fanpageName;

    @JoinColumn(name = "leader_id")
    private Integer leaderId;

    @Column(name = "status")
    @Enumerated(EnumType.STRING)
    private Fanpage.STATUS status;

    @Column(name = "create_time")
    private Long createTime;

    @Column(name = "subscriber")
    private Integer subscriber;

    @Column(name = "cover_image")
    private String cover_image;

    @Column(name = "description")
    private String description;

    @Column(name = "avatar_image")
    private String avatar_image;

    public void setCoverImage(String url) {
        this.cover_image = url;
    }

    public enum STATUS{
        VERIFIED,
        NOT_VERIFY
    }
}

File: front-end/lib/models/entity/fanpage.dart
import 'package:json_annotation/json_annotation.dart';

part 'fanpage.g.dart';

@JsonSerializable()
class Fanpage {
  int? id;
  String? fanpageName;
  String? description;
  int? leaderId;
  FanpageStatus? status;
  @JsonKey(fromJson: _fromJson, toJson: _toJson)
  DateTime? createTime;
  int? subscriber;
  @JsonKey(name: 'cover_image')
  String? coverImage;
  @JsonKey(name: 'avatar_image')
  String? avatarImage;

  Fanpage({this.id, this.fanpageName, this.description, this.leaderId, this.status, this.createTime, this.subscriber, this.coverImage, this.avatarImage});
  factory Fanpage.fromJson(Map<String, dynamic> json) => _$FanpageFromJson(json);

  Map<String, dynamic> toJson() => _$FanpageToJson(this);

  static DateTime? _fromJson(int? int) => int != null ? DateTime.fromMillisecondsSinceEpoch(int) : null;
  static int? _toJson(DateTime? time) => time?.millisecondsSinceEpoch;
}

enum FanpageStatus {
  VERIFIED,
  NOT_VERIFY,
}
